castoredc_api/client/castoredc_api_client.py

```diff
@@ -849,8 +1020,20 @@ def retrieve_rest_of_pages(self, url, params, pages):
                 {"page": str(page), "page_size": "1000", **params}
                 for page in range(2, pages)
             ]
-        responses = asyncio.run(self.async_get(url=url, params=params))
-        return [self.handle_response(response) for response in responses]
+        try:
+            # Test if there is a running event loop
+            # If there is, we can't use async code
+            # Solution for IPython consoles (Jupiter Notebooks, Spyder3)
+            asyncio.get_running_loop()
+            responses = [
+                self.sync_get(url, param) for param in tqdm(params, file=sys.stdout)
+            ]
```

Comment:
If there is, why not schedule this onto the already running event loop, instead of falling back to synchronous code?

(Hint: [asyncio.create_task](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task))

Reply:
Because I couldn't (and can't) seem to figure out how to await the coroutine that returns the responses.

For example
```python
        try:
            # Running eventloop, can't use async code
            loop = asyncio.get_running_loop()
            task = self.async_get(url=url, params=params)
            loop.create_task(task)
            responses = [self.handle_response(response) for response in task]
```

Will give an error as the task is not awaited and the sync code tries to iterate over the coroutine instead of the responses returned by the coroutine.
Moreover, we can't use async in this part of the code because it is sync code.
Maybe I am missing something, but the internet/Google isn't helping me either.

I found this workaround; https://pypi.org/project/nest-asyncio/ but I'd rather not rely on such patches.
Do you know if there is any way to await asyncio.create_task in sync code?

Reply:
Right, I understand your problem. It's difficult to wait on tasks and coroutines from outside of async code.

I'm fairly sure [asyncio.run_coroutine_threadsafe](https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe) is the way to go here. Here's an example: https://gist.github.com/zoni/6cc98558c3be230714b209477b23f0e3

Does that help?

Reply:
Nope, sadly this results in a timeout error when specifying a timeout in future.result or just nothing happening when not specifying a TimeOut. I will open an issue for a feature update because nested async calls is a problem according to a lot of resources (Google, StackOverflow)